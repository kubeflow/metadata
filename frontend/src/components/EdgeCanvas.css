.edgeCanvas {
  position: absolute;
  width: 160px;
  height: 440px;
  border: 0;
  display: inline-block;
  margin-left: 260px;
  margin-top: 74px;
  z-index: -1;
}
.edgeCanvas.reverse {
  margin-left: 0;
  transform: translateX(-100%);
}

.kWKfgJ {
  display: block;
  overflow: visible;
  position: absolute;
}