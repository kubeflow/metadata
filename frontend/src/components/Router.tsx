/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Banner, {BannerProps} from '../components/Banner';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import Page404 from '../pages/404';
import Snackbar, {SnackbarProps} from '@material-ui/core/Snackbar';
import Toolbar, {ToolbarProps} from './Toolbar';
import {Route, Switch, Redirect, HashRouter} from 'react-router-dom';
import {classes, stylesheet} from 'typestyle';
import {commonCss} from '../Css';
import ArtifactList from '../pages/ArtifactList';
import ArtifactDetails from '../pages/ArtifactDetails';
import ExecutionList from '../pages/ExecutionList';
import SideNav from '../pages/SideNav';

const css = stylesheet({
  dialog: {
    minWidth: 250,
  },
});

export enum RouteParams {
  ARTIFACT_TYPE = 'artifactType',
  EXECUTION_TYPE = 'executionType',
  ID = 'id',
}

export const RoutePage = {
  ARTIFACTS: '/artifacts',
  ARTIFACT_DETAILS: `/artifact_types/:${RouteParams.ARTIFACT_TYPE}+/artifacts/:${RouteParams.ID}`,
  EXECUTIONS: '/executions',
  EXECUTION_DETAILS: `/execution_types/:${RouteParams.EXECUTION_TYPE}+/executions/:${RouteParams.ID}`,
};

export interface DialogProps {
  buttons?: Array<{onClick?: () => any, text: string}>;
  // TODO: This should be generalized to any react component.
  content?: string;
  onClose?: () => any;
  open?: boolean;
  title?: string;
}

interface RouteConfig {
  path: string,
  Component: React.ComponentClass,
  view?: any
}

interface RouteComponentState {
  bannerProps: BannerProps;
  dialogProps: DialogProps;
  snackbarProps: SnackbarProps;
  toolbarProps: ToolbarProps;
}

class Router extends React.Component<{}, RouteComponentState> {

  private routes: RouteConfig[] = [
    {path: RoutePage.ARTIFACTS, Component: ArtifactList},
    {path: RoutePage.ARTIFACT_DETAILS, Component: ArtifactDetails},
    {path: RoutePage.EXECUTIONS, Component: ExecutionList},
    {path: RoutePage.EXECUTION_DETAILS, Component: ArtifactDetails},
  ];

  constructor(props: any) {
    super(props);
    this.state = {
      bannerProps: {},
      dialogProps: {open: false},
      snackbarProps: {autoHideDuration: 5000, open: false},
      toolbarProps: {breadcrumbs: [{displayName: '', href: ''}], actions: [], ...props},
    };
  }

  public render(): JSX.Element {
    const childProps = {
      toolbarProps: this.state.toolbarProps,
      updateBanner: this._updateBanner.bind(this),
      updateDialog: this._updateDialog.bind(this),
      updateSnackbar: this._updateSnackbar.bind(this),
      updateToolbar: this._updateToolbar.bind(this),
    };
    return (
      <HashRouter>
        <div className={commonCss.page}>
          <div className={commonCss.flexGrow}>
          <Route render={({ ...props }) => (<SideNav page={props.location.pathname} {...props} />)} />
            <div className={classes(commonCss.page)}>
              <Route render={({...props}) => (<Toolbar {...this.state.toolbarProps} {...props} />)} />
              {this.state.bannerProps.message
                && <Banner
                  message={this.state.bannerProps.message}
                  mode={this.state.bannerProps.mode}
                  additionalInfo={this.state.bannerProps.additionalInfo}
                  refresh={this.state.bannerProps.refresh} />}
              <Switch>
                <Route exact={true} path={'/'} render={({...props}) => (
                  <Redirect to={RoutePage.ARTIFACTS} {...props} />
                )} />
                {this.routes.map((route, i) => {
                  const {path, Component, ...otherProps} = {...route};
                  return <Route key={i} exact={true} path={path} render={({...props}) => (
                    <Component {...props} {...childProps} {...otherProps} />
                  )} />;
                })}

                {/* 404 */}
                {<Route render={({...props}) => <Page404 {...props} {...childProps} />} />}
              </Switch>

              <Snackbar
                autoHideDuration={this.state.snackbarProps.autoHideDuration}
                message={this.state.snackbarProps.message}
                open={this.state.snackbarProps.open}
                onClose={this._handleSnackbarClose.bind(this)}
              />
            </div>
          </div>

          <Dialog open={this.state.dialogProps.open !== false} classes={{paper: css.dialog}}
            className='dialog' onClose={() => this._handleDialogClosed()}>
            {this.state.dialogProps.title && (
              <DialogTitle> {this.state.dialogProps.title}</DialogTitle>
            )}
            {this.state.dialogProps.content && (
              <DialogContent className={commonCss.prewrap}>
                {this.state.dialogProps.content}
              </DialogContent>
            )}
            {this.state.dialogProps.buttons && (
              <DialogActions>
                {this.state.dialogProps.buttons.map((b, i) =>
                  <Button key={i} onClick={() => this._handleDialogClosed(b.onClick)}
                    className='dialogButton' color='secondary'>
                    {b.text}
                  </Button>)}
              </DialogActions>
            )}
          </Dialog>
        </div>
      </HashRouter>
    );
  }

  private _updateDialog(dialogProps: DialogProps): void {
    // Assuming components will want to open the dialog by defaut.
    if (dialogProps.open === undefined) {
      dialogProps.open = true;
    }
    this.setState({dialogProps});
  }

  private _handleDialogClosed(onClick?: () => void): void {
    this.setState({dialogProps: {open: false}});
    if (onClick) {
      onClick();
    }
    if (this.state.dialogProps.onClose) {
      this.state.dialogProps.onClose();
    }
  }

  private _updateToolbar(newToolbarProps: Partial<ToolbarProps>): void {
    const toolbarProps = Object.assign(this.state.toolbarProps, newToolbarProps);
    this.setState({toolbarProps});
  }

  private _updateBanner(bannerProps: BannerProps): void {
    this.setState({bannerProps});
  }

  private _updateSnackbar(snackbarProps: SnackbarProps): void {
    snackbarProps.autoHideDuration =
      snackbarProps.autoHideDuration || this.state.snackbarProps.autoHideDuration;
    this.setState({snackbarProps});
  }

  private _handleSnackbarClose(): void {
    this.setState({snackbarProps: {open: false, message: ''}});
  }
}

// TODO: loading/error experience until backend is reachable

export default Router;
