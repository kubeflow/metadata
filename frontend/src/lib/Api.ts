import {MetadataServiceApi} from '../apis/service';
import {
  MetadataStoreServiceClient,
  ServiceError, UnaryResponse
} from "../generated/src/apis/metadata/metadata_store_service_pb_service";

/** Known Artifact properties */
export enum ArtifactProperties {
  ALL_META = '__ALL_META__',
  CREATE_TIME = 'create_time',
  DESCRIPTION = 'description',
  NAME = 'name',
  PIPELINE_NAME = 'pipeline_name',
  VERSION = 'version',
}

/** Known Artifact custom properties */
export enum ArtifactCustomProperties {
  WORKSPACE = '__kf_workspace__',
  RUN = '__kf_run__',
}

/** Known Execution properties */
export enum ExecutionProperties {
  NAME = 'name',
  PIPELINE_NAME = 'pipeline_name',
  STATE = 'state',
}

/** Known Execution custom properties */
export enum ExecutionCustomProperties {
  WORKSPACE = '__kf_workspace__',
}

/** Format for a List request */
export interface ListRequest {
  filter?: string;
  orderAscending?: boolean;
  pageSize?: number;
  pageToken?: string;
  sortBy?: string;
}

type Callback<R> = (err: ServiceError | null, res: R | null) => void;
type MetadataApiMethod<T, R> = (request: T, callback: Callback<R>) => UnaryResponse;
type PromiseBasedMetadataApiMethod<T, R> = (request: T) => Promise<{ response: R | null, error: ServiceError | null }>;

/**
 * Converts a callback based api method to promise based api method.
 */
function makePromiseApi<T, R>(apiMethod: MetadataApiMethod<T, R>): PromiseBasedMetadataApiMethod<T, R> {
  return (request: T) => new Promise((resolve, reject) => {
    const handler = (error: ServiceError | null, response: R | null) => {
      // resolve both response and error to keep type information
      resolve({ response, error });
    };
    apiMethod(request, handler);
  });
}

const metadataServiceClient = new MetadataStoreServiceClient('');

// TODO: add all other api methods we need here.
const metadataServicePromiseClient = {
  getArtifactTypes: makePromiseApi(metadataServiceClient.getArtifactTypes.bind(metadataServiceClient)),
  getArtifactsByID: makePromiseApi(metadataServiceClient.getArtifactsByID.bind(metadataServiceClient)),
  getEventsByArtifactIDs: makePromiseApi(metadataServiceClient.getEventsByArtifactIDs.bind(metadataServiceClient)),
  getEventsByExecutionIDs: makePromiseApi(metadataServiceClient.getEventsByExecutionIDs.bind(metadataServiceClient)),
  getExecutionsByID: makePromiseApi(metadataServiceClient.getExecutionsByID.bind(metadataServiceClient)),
};

/**
 * Class to wrap backend APIs.
 */
export class Api {

  private static instance: Api;
  private metadataServiceApi: MetadataServiceApi;

  /**
   * Factory function to return an Api instance.
   */
  public static getInstance(): Api {
    if (!Api.instance) {
      const location = window.location;
      let path = `${location.protocol}//${location.host}${location.pathname}`;
      if (path.endsWith('/')) path = path.slice(0, path.length - 1);
      Api.instance = new Api(path);
    }
    return Api.instance;
  }

  private constructor(basePath: string) {
    this.metadataServiceApi = new MetadataServiceApi({basePath});
  }

  get metadataService(): MetadataServiceApi {
    return this.metadataServiceApi;
  }

  get metadataStoreService() {
    return metadataServicePromiseClient;
  }
}
