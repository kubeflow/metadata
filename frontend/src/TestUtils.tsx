/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable */
// tslint:disable: object-literal-sort-keys

// Because this is test utils.

import * as React from 'react';
// @ts-ignore
import createRouterContext from 'react-router-test-context';
import {PageProps, Page} from './pages/Page';
import {ToolbarActionConfig} from './components/Toolbar';
import {match} from 'react-router';
import {mount, ReactWrapper} from 'enzyme';
import {object} from 'prop-types';
import {format} from 'prettier';
import snapshotDiff from 'snapshot-diff';
import {
  Artifact,
  ArtifactProperties,
  ArtifactCustomProperties,
  Value,
} from '@kubeflow/frontend';

export default class TestUtils {
  /**
   * Mounts the given component with a fake router and returns the mounted tree
   */
  // tslint:disable-next-line:variable-name
  public static mountWithRouter(component: React.ReactElement<any>): ReactWrapper {
    const childContextTypes = {
      router: object,
    };
    const context = createRouterContext();
    const tree = mount(component, { context, childContextTypes });
    return tree;
  }

  /**
   * Flushes all already queued promises and returns a promise. Note this will
   * only work if the promises have already been queued, so it cannot be used to
   * wait on a promise that hasn't been dispatched yet.
   */
  public static flushPromises(): Promise<void> {
    return new Promise(resolve => setImmediate(resolve));
  }

  /**
   * Adds a one-time mock implementation to the provided spy that mimics an error
   * network response
   */
  public static makeErrorResponseOnce(spy: jest.MockInstance<unknown, unknown[]>, message: string): void {
    spy.mockImplementationOnce(() => {
      throw {
        text: () => Promise.resolve(message),
      };
    });
  }

  /**
   * Generates a customizable PageProps object that can be passed to initialize
   * Page components, taking care of setting ToolbarProps properly, which have
   * to be set after component initialization.
   */
  // tslint:disable-next-line:variable-name
  public static generatePageProps(
    PageElement: new (_: PageProps) => Page<any, any>,
    location: Location,
    matchValue: match,
    historyPushSpy: jest.SpyInstance<unknown> | null,
    updateBannerSpy: jest.SpyInstance<unknown> | null,
    updateDialogSpy: jest.SpyInstance<unknown> | null,
    updateToolbarSpy: jest.SpyInstance<unknown> | null,
    updateSnackbarSpy: jest.SpyInstance<unknown> | null,
  ): PageProps {
    const pageProps = {
      history: { push: historyPushSpy } as any,
      location: location as any,
      match: matchValue,
      toolbarProps: { actions: {}, breadcrumbs: [], pageTitle: '' },
      updateBanner: updateBannerSpy as any,
      updateDialog: updateDialogSpy as any,
      updateSnackbar: updateSnackbarSpy as any,
      updateToolbar: updateToolbarSpy as any,
    } as PageProps;
    pageProps.toolbarProps = new PageElement(pageProps).getInitialToolbarState();
    // The toolbar spy gets called in the getInitialToolbarState method, reset it
    // in order to simplify tests
    if (updateToolbarSpy) {
      updateToolbarSpy.mockReset();
    }
    return pageProps;
  }

  public static getToolbarButton(
    updateToolbarSpy: jest.SpyInstance<unknown>,
    buttonKey: string,
  ): ToolbarActionConfig {
    const lastCallIdx = updateToolbarSpy.mock.calls.length - 1;
    const lastCall = updateToolbarSpy.mock.calls[lastCallIdx][0];
    return lastCall.actions[buttonKey];
  }
}

/**
 * Generate diff text for two HTML strings.
 * Recommend providing base and update annotations to clarify context in the diff directly.
 */
export function diffHTML({
  base,
  update,
  baseAnnotation,
  updateAnnotation,
}: {
  base: string;
  baseAnnotation?: string;
  update: string;
  updateAnnotation?: string;
}) {
  return diff({
    base: formatHTML(base),
    update: formatHTML(update),
    baseAnnotation,
    updateAnnotation,
  });
}

export function diff({
  base,
  update,
  baseAnnotation,
  updateAnnotation,
}: {
  base: string;
  baseAnnotation?: string;
  update: string;
  updateAnnotation?: string;
}) {
  return snapshotDiff(base, update, {
    stablePatchmarks: true, // Avoid line numbers in diff, so that diffs are stable against irrelevant changes
    aAnnotation: baseAnnotation,
    bAnnotation: updateAnnotation,
  });
}

function formatHTML(html: string): string {
  return format(html, { parser: 'html' });
}

/**	
 * Generates a customizable PageProps object that can be passed to initialize	
 * Page components, taking care of setting ToolbarProps properly, which have	
 * to be set after component initialization.	
 */	
// tslint:disable-next-line:variable-name	
export function generatePageProps(	
  // tslint:disable-next-line: variable-name
  PageElement: new (_: PageProps) => Page<any, any>, location: Location,	
  matchValue: match, historyPushSpy: jest.SpyInstance | null,	
  updateBannerSpy: jest.SpyInstance | null,	
  updateDialogSpy: jest.SpyInstance | null,	
  updateToolbarSpy: jest.SpyInstance | null,	
  updateSnackbarSpy: jest.SpyInstance | null): PageProps {	
  const pageProps = {	
    history: {push: historyPushSpy} as any,	
    location: location as any,	
    match: matchValue,	
    toolbarProps: {actions: [], breadcrumbs: [], pageTitle: ''},	
    updateBanner: updateBannerSpy as any,	
    updateDialog: updateDialogSpy as any,	
    updateSnackbar: updateSnackbarSpy as any,	
    updateToolbar: updateToolbarSpy as any,	
  } as unknown as PageProps;	
  pageProps.toolbarProps = new PageElement(pageProps).getInitialToolbarState();	
  // The toolbar spy gets called in the getInitialToolbarState method, reset	
  // it in order to simplify tests	
  if (updateToolbarSpy) {	
    updateToolbarSpy.mockReset();	
  }	
  return pageProps;	
}	

export function getToolbarButton(	
  updateToolbarSpy: jest.SpyInstance, title: string): ToolbarActionConfig {	
  const lastCallIdx = updateToolbarSpy.mock.calls.length - 1;	
  const lastCall = updateToolbarSpy.mock.calls[lastCallIdx][0];	
  return lastCall.actions.find((b: any) => b.title === title);	
}	

export const doubleValue = (number: number) => {	
  const value = new Value();	
  value.setDoubleValue(number);	
  return value;	
};	

export const intValue = (number: number) => {	
  const value = new Value();	
  value.setIntValue(number);	
  return value;	
};	

export const stringValue = (string: String) => {	
  const value = new Value();	
  value.setStringValue(String(string));	
  return value;	
};	

export const buildTestModel = () => {	
  const model = new Artifact();	
  model.setId(1);	
  model.setTypeId(1);	
  model.setUri('gs://my-bucket/mnist');	
  model.getPropertiesMap().set(ArtifactProperties.NAME, stringValue('test model'));	
  model.getPropertiesMap().set(ArtifactProperties.DESCRIPTION, stringValue('A really great model'));	
  model.getPropertiesMap().set(ArtifactProperties.VERSION, stringValue('v1'));	
  model.getPropertiesMap().set(ArtifactProperties.CREATE_TIME, stringValue('2019-06-12T01:21:48.259263Z'));	
  model.getPropertiesMap().set(ArtifactProperties.ALL_META, stringValue(	
      '{"hyperparameters": {"early_stop": true, ' +	
      '"layers": [10, 3, 1], "learning_rate": 0.5}, ' +	
      '"model_type": "neural network", ' +	
      '"training_framework": {"name": "tensorflow", "version": "v1.0"}}'));	
  model.getCustomPropertiesMap().set(ArtifactCustomProperties.WORKSPACE, stringValue('workspace-1'));	
  model.getCustomPropertiesMap().set(ArtifactCustomProperties.RUN, stringValue('1'));	
  return model	
};	

export const testModel = buildTestModel();