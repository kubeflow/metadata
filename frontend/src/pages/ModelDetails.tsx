/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import {Page} from './Page';
import {ModelInfo, ModelInfoProps} from '../components/ModelInfo';
import {ToolbarProps} from '../components/Toolbar';
import {RoutePage, RouteParams} from '../components/Router';
import {Api, CustomProperties} from '../lib/Api';
import {MlMetadataArtifact} from '../apis/service';
import {classes} from 'typestyle';
import {commonCss, padding} from '../Css';
import {LinearProgress} from '@material-ui/core';
import {logger} from '../lib/Utils';

interface ModelDetailsState {
  model?: MlMetadataArtifact;
}

// From ../../schema/alpha/artifacts/model.json
interface ModelSchema {
  hyperparameters?: {[name: string]: string};
  model_type?: string;
  training_framework?: {name: string, version: string};
}

export default class ModelDetails extends Page<{}, ModelDetailsState> {
  private api = Api.getInstance();

  constructor(props: {}) {
    super(props);
    this.state = {};
  }

  public async componentDidMount(): Promise<void> {
    return this.load();
  }

  public render(): JSX.Element {
    const {model} = this.state;
    if (!model) return <LinearProgress />;
    const modelInfoProps: ModelInfoProps = {
      description: model.properties!.description.string_value,
      version: model!.properties!.version.string_value,
      createTime: model!.properties!.create_time.string_value,
      uri: model!.uri,
      workspace: model!.custom_properties![CustomProperties.WORKSPACE].string_value,
      run: model!.custom_properties![CustomProperties.RUN].string_value
    };
    // let parsedModelInfo: ModelInfoProps
    let parsedMetadata: ModelSchema;
    if (model && model.properties![CustomProperties.ALL_META]) {
      try {
        parsedMetadata = JSON.parse(
          model.properties![CustomProperties.ALL_META]!.string_value!);
        modelInfoProps.hyperparameters = parsedMetadata.hyperparameters;
        modelInfoProps.modelType = parsedMetadata.model_type;
        modelInfoProps.trainingFramework = parsedMetadata.training_framework;
      } catch (err) {
        logger.error('Unable to parse CustomProperties.ALL_META', err);
      }
    }
    return (
      <div className={classes(commonCss.page, padding(20, 'lr'))}>
        {!!model && (
          <div>
            <ModelInfo {...modelInfoProps} />
            <pre style={{marginTop: '10px', borderTop: '1px solid #000'}}>
              {JSON.stringify(this.state.model, undefined, 2)}
            </pre>
          </div>
        )}
      </div>
    );
  }

  public getInitialToolbarState(): ToolbarProps {
    return {
      actions: [],
      breadcrumbs: [{displayName: 'Artifacts', href: RoutePage.ARTIFACTS}],
      pageTitle: `Model ${this.props.match.params[RouteParams.artifactId]} details`
    };
  }

  public async refresh(): Promise<void> {
    return this.load();
  }

  private async load(): Promise<void> {
    // TODO: Use the getArtifact API method which doesn't seem to be generated
    const id = this.props.match.params[RouteParams.artifactId];
    try {
      const response = await this.api.metadataService.listArtifacts2();
      const [model] = response.artifacts!.filter((a) => id === a.id);
      if (model) {
        let title = model.properties!.name!.string_value!;
        if (model.properties!.version) {
          title += ` (version: ${model.properties!.version!.string_value})`;
        }
        this.props.updateToolbar({
          pageTitle: title
        });
        this.setState({model});
      }
    } catch (err) {
      this.showPageError(`Unable to retrieve Model ${id}`, err);
    }
  }
}
