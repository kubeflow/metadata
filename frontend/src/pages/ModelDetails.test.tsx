import * as React from 'react';
import ModelDetails from './ModelDetails';
import {PageProps} from './Page';
import {shallow, ShallowWrapper, ReactWrapper} from 'enzyme';
import {Api} from '../lib/Api';
import * as TestUtils from '../TestUtils'
import {RouteParams} from '../components/Router';
import {ApiListArtifactsResponse} from '../apis/service';

describe('ModelDetails', () => {
  let tree: ShallowWrapper | ReactWrapper;
  const updateBannerSpy = jest.fn();
  const updateToolbarSpy = jest.fn();
  const historyPushSpy = jest.fn();
  const mockListArtifacts = jest.spyOn(
    Api.getInstance().metadataService, 'listArtifacts2');
  const fakeArtifactsResponse: ApiListArtifactsResponse = {
    artifacts: [
      {
        id: "1",
        type_id: "1",
        uri: "gs://my-bucket/mnist",
        properties: {
          name: {string_value: 'model'},
          description: {string_value: 'A really great model'},
          version: {string_value: 'v1'},
          create_time: {string_value: '2019-06-12T01:21:48.259263Z'},
          __ALL_META__: {
            string_value: '{"hyperparameters": {"early_stop": true, ' +
              '"layers": [10, 3, 1], "learning_rate": 0.5}, ' +
              '"model_type": "neural network", ' +
              '"training_framework": {"name": "tensorflow", "version": "v1.0"}}'
          }
        },
        custom_properties: {
          __kf_workspace__: {string_value: 'workspace-1'},
          __kf_run__: {string_value: '1'},
        },
      },
      {
        id: "2",
        type_id: "3",
        uri: "gs://my-bucket/dataset2",
        properties: {
          name: {string_value: 'dataset'},
          description: {string_value: 'A really great dataset'},
          version: {string_value: 'v2'},
          create_time: {string_value: '2019-06-16T01:21:48.259263Z'},
        },
        custom_properties: {
          __kf_workspace__: {string_value: 'workspace-1'},
          __kf_run__: {string_value: '1'},
        },
      }
    ]
  };
  const FAKE_MODEL_ID = '1';

  function generateProps(): PageProps {
    return TestUtils.generatePageProps(
      ModelDetails,
      {} as any,
      {params: {[RouteParams.artifactId]: FAKE_MODEL_ID}} as any,
      historyPushSpy,
      updateBannerSpy,
      jest.fn(),
      updateToolbarSpy,
      jest.fn());
  }

  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => tree.unmount());

  it('Renders progress bar with no data', () => {
    tree = shallow(<ModelDetails {...generateProps()} />);

    expect(tree).toMatchSnapshot();
  });

  it('Renders with a Model and updates the page title', async () => {
    mockListArtifacts.mockResolvedValue(fakeArtifactsResponse);
    tree = TestUtils.mountWithRouter(<ModelDetails {...generateProps()} />);

    await mockListArtifacts;
    await TestUtils.flushPromises();
    tree.update();
    expect(tree).toMatchSnapshot();
    expect(updateToolbarSpy).toHaveBeenLastCalledWith(expect.objectContaining({
      pageTitle: 'model (version: v1)',
    }));
  });

  it('Shows error when Models cannot be retrieved', async () => {
    mockListArtifacts.mockRejectedValue(new Error('Get Model error'));
    tree = TestUtils.mountWithRouter(<ModelDetails {...generateProps()} />);

    await mockListArtifacts;
    await TestUtils.flushPromises();
    expect(updateBannerSpy).toHaveBeenCalledWith(expect.objectContaining({
      message: 'Unable to retrieve Model 1. ' +
        'Click Details for more information.',
      mode: 'error',
    }));
  });

  // TODO: This can be removed once the getArtifact API method is in place
  it('Shows error when specified Model cannot be found', async () => {
    mockListArtifacts.mockResolvedValue(fakeArtifactsResponse);
    const props = generateProps();
    props.match = {params: {[RouteParams.artifactId]: '3'}} as any;
    tree = TestUtils.mountWithRouter(<ModelDetails {...props} />);

    await mockListArtifacts;
    await TestUtils.flushPromises();
    expect(updateBannerSpy).toHaveBeenCalledWith(expect.objectContaining({
      message: 'Unable to retrieve Model 3. ' +
        'Click Details for more information.',
      mode: 'error',
    }));
  });
});
