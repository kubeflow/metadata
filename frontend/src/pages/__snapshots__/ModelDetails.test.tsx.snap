// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelDetails Renders progress bar with no data 1`] = `<WithStyles(LinearProgress) />`;

exports[`ModelDetails Renders with a Model and updates the page title 1`] = `
<ModelDetails
  history={
    Object {
      "push": [MockFunction],
    }
  }
  location={Object {}}
  match={
    Object {
      "params": Object {
        "id": "1",
      },
    }
  }
  toolbarProps={
    Object {
      "actions": Array [],
      "breadcrumbs": Array [
        Object {
          "displayName": "Artifacts",
          "href": "/artifacts",
        },
      ],
      "pageTitle": "Model 1 details",
    }
  }
  updateBanner={[MockFunction]}
  updateDialog={[MockFunction]}
  updateSnackbar={[MockFunction]}
  updateToolbar={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "actions": Array [],
            "breadcrumbs": Array [
              Object {
                "displayName": "Artifacts",
                "href": "/artifacts",
              },
            ],
            "pageTitle": "Model 1 details",
          },
        ],
        Array [
          Object {
            "pageTitle": "model (version: v1)",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="page_f1flacxk f9qqjqd"
  >
    <ModelInfo
      createTime="2019-06-12T01:21:48.259263Z"
      description="A really great model"
      hyperparameters={
        Object {
          "early_stop": true,
          "layers": Array [
            10,
            3,
            1,
          ],
          "learning_rate": 0.5,
        }
      }
      modelType="neural network"
      run="1"
      trainingFramework={
        Object {
          "name": "tensorflow",
          "version": "v1.0",
        }
      }
      uri="gs://my-bucket/mnist"
      version="v1"
      workspace="workspace-1"
    >
      <section>
        <h1
          className="header_f189n3y"
        >
          Model info
        </h1>
        <dl
          className="modelInfo_fnuket9"
        >
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Type
            </dt>
            <dd
              className="value_f1ej2mlj"
            >
              neural network
            </dd>
          </div>
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Description
            </dt>
            <dd
              className="value_f1ej2mlj"
            >
              A really great model
            </dd>
          </div>
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Version ID
            </dt>
            <dd
              className="value_f1ej2mlj"
            >
              v1
            </dd>
          </div>
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Workspace
            </dt>
            <dd
              className="value_f1ej2mlj"
            >
              workspace-1
            </dd>
          </div>
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Run
            </dt>
            <dd
              className="value_f1ej2mlj"
            >
              1
            </dd>
          </div>
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Creation time
            </dt>
            <dd
              className="value_f1ej2mlj"
            >
              6/11/2019, 6:21:48 PM
            </dd>
          </div>
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Hyperparameters
            </dt>
            <dd
              className="value_f1ej2mlj flexColumn_fr3ikkd"
            >
              <span
                key="0"
              >
                early_stop: true
              </span>
              <span
                key="1"
              >
                layers: 10,3,1
              </span>
              <span
                key="2"
              >
                learning_rate: 0.5
              </span>
            </dd>
          </div>
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Training data
            </dt>
            <dd
              className="value_f1ej2mlj"
            >
              gs://my-bucket/mnist
            </dd>
          </div>
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Training framework
            </dt>
            <dd
              className="value_f1ej2mlj"
            >
              tensorflow
                   v1.0
            </dd>
          </div>
          <div
            className="field_f16919hf"
          >
            <dt
              className="term_f1wopveh"
            >
              Training data
            </dt>
            <dd
              className="value_f1ej2mlj"
            >
              gs://my-bucket/mnist
            </dd>
          </div>
        </dl>
      </section>
    </ModelInfo>
  </div>
</ModelDetails>
`;
